### ISSUE TITLE
[BUG/Integrity] Silent State-Dict Key Dropping, Unchecked Zero-Padded Weights,

### ISSUE THREAD

**Issue by Dikshant-Neupane** [microsoft/Mage#21]
## Summary
A deep-dive technical audit of `Mage` core utilities and pipeline loading routines (`mage_flow.py`, `utils.py`, `mage_text.py`, and `pipeline.py`) identified several critical integration guards and checkpoint loading flaws. 

These issues allow silent weight corruption, partial zero-initialization of projection layers, and hidden execution failures during model state restoration.

---

### 1.1 Silent Key Dropping in `load_state_dict(strict=False)`
* **Location:** `mage_flow.py` & `pipeline.py`
* **Root Cause:** Missing checkpoint keys are suppressed or logged as non-fatal warnings when loading weights with `strict=False, assign=True`.
* **Impact:** Mismatched architecture checkpoints (e.g., attempting to load a 1.3B checkpoint into a 4B model backbone) run silently with uninitialized layers instead of throwing an explicit error.

### 1.2 Unchecked Zero-Padding on Parameter Shape Mismatches
* **Location:** `utils.py:optionally_expand_state_dict()`
* **Root Cause:** When checkpoint shape dimensions do not match model expectations, the function slices parameter weights into zero-initialized tensors without validating if the expansion is safe.
* **Impact:** Critical projection layers (`img_in.weight`, `txt_in.weight`) become partially populated with zeros, causing dead channels during inference without raising errors.

### 1.3 Unhandled Exceptions in Model Mode Restoration
* **Location:** `mage_text.py:_full_output_mode()`
* **Root Cause:** Context manager exceptions during output mode restoration are silently caught with `except Exception: pass`.
* **Impact:** If restoration fails, the text model remains stuck in an incorrect mode, causing subsequent prompt screenings or embeddings to evaluate under wrong assumptions.

### 1.4 Missing VAE Latent vs. Transformer Channel Assertions
* **Location:** `pipeline.py:compute_vae_encodings()`
* **Root Cause:** The pipeline lacks a runtime assertion check verifying that `vae.latent_channels == transformer.in_channels`.
* **Impact:** Configuration mismatches silently truncate or slice incoming latent tensors prior to linear projection, leading to degraded spatial outputs.

---

## 2. Proposed Solutions & Refactored Code

```python
import logging
import torch
import torch.nn as nn

logger = logging.getLogger("MageIntegrityFixes")

# Solution 1 & 2: Strict Checkpoint Validation Guard
def validate_and_load_state_dict(model: nn.Module, state_dict: dict, strict_critical: bool = True):
    """Ensures critical projection layers are not silently dropped or zero-padded."""
    CRITICAL_LAYERS = {"img_in.weight", "txt_in.weight", "proj_out.weight", "final_layer.linear.weight"}
    model_state = model.state_dict()
    missing_critical = []
    mismatched_shapes = []

    for name in CRITICAL_LAYERS:
        if name in model_state and name not in state_dict:
            missing_critical.append(name)
        elif name in state_dict and state_dict[name].shape != model_state[name].shape:
            mismatched_shapes.append((name, state_dict[name].shape, model_state[name].shape))

    if missing_critical and strict_critical:
        raise KeyError(f"Critical layers missing from checkpoint: {missing_critical}")
        
    if mismatched_shapes:
        raise ValueError(f"Shape mismatches in critical layers: {mismatched_shapes}")

    # Proceed with safe loading
    return model.load_state_dict(state_dict, strict=False)

# Solution 3: Explicit Error Handling for Mode Restoration
class RefactoredFullOutputMode:
    def __init__(self, model):
        self.model = model
        self._prev_mode = None

    def __enter__(self):
        self._prev_mode = getattr(self.model, "output_mode", None)
        self.model.output_mode = "full"
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        try:
            self.model.output_mode = self._prev_mode
        except Exception as exc:
            logger.error(f"Failed to restore text model output mode to '{self._prev_mode}': {exc}")
            # Re-raise to prevent executing subsequent steps in an invalid state
            raise exc

# Solution 4: Pipeline Architectural Channel Validation
def validate_pipeline_config(vae: nn.Module, transformer: nn.Module):
    vae_channels = getattr(vae, "latent_channels", None)
    transformer_channels = getattr(transformer, "in_channels", None)
    
    if vae_channels is not None and transformer_channels is not None:
        if vae_channels != transformer_channels:
            raise ValueError(
                f"Pipeline Dimension Mismatch: VAE output latent channels ({vae_channels}) "
                f"do not match Transformer input channels ({transformer_channels})."
            )
